// state-pool/esm/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useLocalState = exports.useGlobalState = exports.useGlobalStateReducer = exports.store = exports.GlobalStateStore = exports.createGlobalstate = exports.GlobalState = void 0;
var GlobalState_1 = require("./GlobalState");
Object.defineProperty(exports, "GlobalState", { enumerable: true, get: function () { return GlobalState_1.GlobalState; } });
Object.defineProperty(exports, "createGlobalstate", { enumerable: true, get: function () { return GlobalState_1.createGlobalstate; } });
var GlobalStateStore_1 = require("./GlobalStateStore");
Object.defineProperty(exports, "GlobalStateStore", { enumerable: true, get: function () { return GlobalStateStore_1.GlobalStateStore; } });
Object.defineProperty(exports, "store", { enumerable: true, get: function () { return GlobalStateStore_1.store; } });
var useGlobalStateReducer_1 = require("./useGlobalStateReducer");
Object.defineProperty(exports, "useGlobalStateReducer", { enumerable: true, get: function () { return useGlobalStateReducer_1.useGlobalStateReducer; } });
var useGlobalState_1 = require("./useGlobalState");
Object.defineProperty(exports, "useGlobalState", { enumerable: true, get: function () { return useGlobalState_1.useGlobalState; } });
var useLocalState_1 = require("./useLocalState");
Object.defineProperty(exports, "useLocalState", { enumerable: true, get: function () { return useLocalState_1.useLocalState; } });
